styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,241 @@
+      if (limit == null) {
+        snapshot.limit = Long.MAX_VALUE;
+        // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+      } else {
+        snapshot.limit = limit.getLong(PROPERTY_LIMIT);
+      }
+
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(CompositeFilter.and(
+              PropertyFilter.ge("__key__", counterId + "-0"),
+              PropertyFilter.lt("__key__", counterId + "-" + NUM_SHARDS)))
+          .setOrderBy(OrderBy.asc("__key__"))
+          .setLimit(NUM_SHARDS)
+          .build();
+      final QueryResults<Entity> shards = datastore.run(queryShards);
+      snapshot.shards = new ArrayList<Long>(NUM_SHARDS);
+      int shardCount;
+      for (shardCount = 0; shards.hasNext(); shardCount++) {
+        Entity shard = shards.next();
+        String shardKey = shard.getKey().toString();
+        int shardIndex = Integer.valueOf(shardKey.substring(shardKey.indexOf('-') + 1));
+        snapshot.shards.set(shardIndex, shard.getLong(PROPERTY_SHARD));
+      }
+      if (shardCount < NUM_SHARDS) {
+        // The counter probably has not been initialized (so we have empty QueryResults). Also
+        // possible that a prior initialize() crashed halfway, or we got a partial list of shards in
+        // QueryResults due to eventual consistency. In any case, repeated initialization eventually
+        // creates all NUM_SHARDS shards.
+        initialize(datastore, counterId);
+      }
+
+      snapshot.updatedAt = Instant.now();
+      return snapshot;
+    }
+
+    private boolean isRecent() {
+      return updatedAt.plus(CACHE_EXPIRY_MILLIS, ChronoUnit.MILLIS).isAfter(Instant.now());
+    }
+
+    private long shardCapacity(int shardIndex) {
+      return limit / NUM_SHARDS + (shardIndex < limit % NUM_SHARDS ? 1 : 0);
+    }
+
+    /**
+     * Returns shard index which _likely_ could be successfully updated by delta, according to our
+     * cached view of the state in Datastore.
+     */
+    private int pickShardWithSpareCapacity(long delta) {
+      List<Integer> candidates = new ArrayList<>();
+
+      for (int i = 0; i < shards.size(); i++) {
+        if (shards.get(i) + delta >= 0
+            && shards.get(i) + delta <= shardCapacity(i)) {
+          candidates.add(i);
+        }
+      }
+
+      if (candidates.isEmpty()) {
+        return new Random().nextInt(NUM_SHARDS);
+        // Or return -1 (and use that to abort the transaction early)?
+      } else {
+        return candidates.get(new Random().nextInt(candidates.size()));
+      }
+    }
+  }
+
+  /**
+   * A weakly consistent view of the state in Datastore, refreshed by ShardedCounter on demand.
+   */
+  private Map<String, CounterSnapshot> counterCache;
+
+  public ShardedCounter(Datastore datastore) {
+    this.datastore = Objects.requireNonNull(datastore);
+    this.counterCache = new ConcurrentHashMap<>();
```

Comment:
Use a [guava cache](https://github.com/google/guava/wiki/CachesExplained) with some max size to avoid the risk of unbounded growth and memory leak?

Reply:
+1 to using Guava, but where do you mean there would be unbounded growth? It's growing with the number of distinct resources, and we don't have a deleteCounter yet - did you mean that?